third_party/nginx_latency_status_module/src/ngx_http_latency_status_handler.c

```diff
@@ -0,0 +1,203 @@
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+// SUCH DAMAGE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "ngx_http_latency_status_handler.h"
+#include "ngx_http_latency_storage.h"
+#include "ngx_http_latency_stub_status_module.h"
+
+
+// Write a single latency record to the output buffer in json format.
+static void write_distribution_content(ngx_buf_t *buffer, char name[], ngx_int_t len_dist, latency_stat *latency_record) {
+  u_char const dist_array_start[] = "    \"distribution\": [";
+  u_char const json_sub_var_end[] = "  },\n";
+
+  buffer->last = ngx_sprintf(buffer->last, "  \"%s\":{\n", name);
+
+  buffer->last = ngx_sprintf(buffer->last, "    \"latency_sum\": %uA,\n", latency_record->sum);
+  buffer->last = ngx_sprintf(buffer->last, "    \"request_count\": %uA,\n", latency_record->request_count);
+  buffer->last = ngx_sprintf(buffer->last, "    \"sum_squares\": %uA,\n", latency_record->sum_squares);
+  buffer->last = ngx_cpymem(buffer->last, dist_array_start, sizeof(dist_array_start) - 1);
+  for (int i = 0; i < len_dist - 1; i++) {
+    buffer->last = ngx_sprintf(buffer->last, "%uA, ", latency_record->distribution[i]);
+  }
+  buffer->last = ngx_sprintf(buffer->last, "%uA],\n", latency_record->distribution[len_dist]);
+  buffer->last = ngx_cpymem(buffer->last, json_sub_var_end, sizeof(json_sub_var_end) - 1);
+}
+
+
+// An http request handler that returns a status page including latency
+// distribution stats. The stats are in json format.
+ngx_int_t ngx_http_latency_stub_status_handler(ngx_http_request_t *r)
+{
+  size_t output_size, latency_stat_size;
+  ngx_int_t request_result;
+  ngx_buf_t *buffer;
+  ngx_chain_t out;
+  ngx_http_latency_main_conf_t *main_conf;
+  ngx_int_t dist_len;
+  ngx_atomic_int_t accepted, handled, active, requests, reading, writing, waiting;
+
+  if (!(r->method & (NGX_HTTP_GET|NGX_HTTP_HEAD))) {
+    return NGX_HTTP_NOT_ALLOWED;
+  }
+
+  request_result = ngx_http_discard_request_body(r);
+
+  if (request_result != NGX_OK) {
+    return request_result;
+  }
+
+  r->headers_out.content_type_len = sizeof("application/json") - 1;
+  ngx_str_set(&r->headers_out.content_type, "application/json");
+  r->headers_out.content_type_lowcase = NULL;
+
+  if (r->method == NGX_HTTP_HEAD) {
+    r->headers_out.status = NGX_HTTP_OK;
+
+    request_result = ngx_http_send_header(r);
+
+    if (request_result == NGX_ERROR || request_result > NGX_OK || r->header_only) {
+      return request_result;
+    }
+  }
+
+  main_conf = ngx_http_get_module_main_conf(r, ngx_http_latency_stub_status_module);
+  dist_len = main_conf->max_exponent + 2;
+
+  u_char const json_start[] = "{\n";
```

Comment:
These json generation code is error prone.  Could you link in protobuf code to use protobuf to json translation 
Similar to [here](https://github.com/cloudendpoints/esp/blob/master/src/nginx/status.cc#L157)

Reply:
The question here is whether this is worth converting the whole thing to C++ code since C isn't one of the officially supported languages for protos.

Reply:
I see.  
How about using a separate c++ lib for the json print, so not need to convert all these codes to c++

BTW,  Do you plan use bazel BUILD here or only at ESP repo?

Reply:
So far I've been building and testing locally using the module with vanilla nginx and using the make system that comes with it. But yes I'd need to switch over bazel to include C++ code.

Reply:
Let us do that in phases.  You can do that in subsequent PR.

Reply:
Is there a way to include the tests (later on if needed)?

Reply:
plane nginx doesn't provide a unit testing system, so another advantage of switching to bazel is that it would make adding unit tests easier.

Reply:
Nginx has "t" tests: they are in perl scripts to do e2e tests.  Maybe you can add them here in this repo.   ESP has bazel plugin to support these tests.

Reply:
Maybe I'm not looking at the right part of the nginx documentation, but I haven't been able to find anything about "t" tests. Are they specific to esp? 

The existing appengine nginx module, iap_jwt_verify_nginx, has some unit and e2e tests, but they're in a mix of C++ and python and are built with bazel.

Reply:
It is [here](https://nginx.googlesource.com/nginx-tests/)